homeassistant/components/shelly/utils.py

```diff
@@ -333,8 +333,8 @@ def get_rpc_key_instances(keys_dict: dict[str, Any], key: str) -> list[str]:
     if key in keys_dict:
         return [key]
 
-    if key == "switch" and "cover:0" in keys_dict:
-        key = "cover"
+    # if key == "switch" and "cover:0" in keys_dict:
+    #     key = "cover"
```

Comment:
commented code

Reply:
This code should be there as is needed for some device and works on RL scenario.
I commented it just for a test as based on that code been there or commented, different test fail.

On my understanding this is due to a single MOCK that will represent always all possible keys, but this is far from been a real device representation.

Reply:
This was a bugfix https://github.com/home-assistant/core/pull/71831 it has nothing to do with the way we mock devices, it doesn't even affect anything mentioned in this PR as this is only for creating sensors for the binary & sensor platforms.

Reply:
> This was a bugfix #71831

I know this is a valid code for some devices, I don't want to change or remove it.

> it has nothing to do with the way we mock devices,

Current mock has both `switch` and `cover` keys:

https://github.com/home-assistant/core/blob/dev/tests/components/shelly/conftest.py#L199
https://github.com/home-assistant/core/blob/dev/tests/components/shelly/conftest.py#L203

Real life devices that are in switch mode, only has `switch` key and use a different code path.
This is why i think we should have specific mock for each device.

Reply:
That still doesn't have anything to do with binary sensor and sensor platform failing, they are not accessing this key. Although we now limit the minimum, Gen2 FW to 1.0.0 which remove the `switch` key from the device when it is in cover mode, older firmware did not remove it so it was valid at that time. 

If you think a specific test is failing since it expects only a `cover` or `switch` key lets fix it (better in a separate PR before this PR by deleting the key which is not needed, similar to what we do for light tests:
https://github.com/home-assistant/core/blob/843c84a325983472dd0844a49ca305382521d589/tests/components/shelly/test_light.py#L223-L229

But as I tried to explain multiple times, this is not the issue and sensor tests are not related to this, if they fail it is because this PR breaks something (which I explained what it breaks above).

I will stop now commenting on this PR since I think you are too fixed on how you want to do this change, but still suggest you try to pass the current tests before modifying them.

Reply:
So if I understand correctly the code in this PR is tested on all of Shelly devices with the failed tests and the failures are only due to tests implemented different than the device?

If it is the case what is holding you from making a code quality PR that improve the tests to have a specific mock for each device  **than we can come back to this PR and it tests should pass?**

Reply:
Reposting as it was deleted for no reason.

Your suggestion was not lost, was already applied:

https://github.com/home-assistant/core/blob/09b8e4368f15c1dd8fa890b15ee8e09a28eb324b/homeassistant/components/shelly/switch.py#L197
https://github.com/home-assistant/core/blob/09b8e4368f15c1dd8fa890b15ee8e09a28eb324b/homeassistant/components/shelly/switch.py#L239

I think you are commenting without looking at the code lately.